### PR TITLE
statefulset-configuration-take2

### PR DESCRIPTION
**Pull request by drohnow**
- Added statefulset and values.yaml files.
- Bumped version
- Updated docs & changelog

### CODE REVIEW

File: charts/replicated-library/README_CHANGELOG.md.gotmpl [hunk 1]
```diff
@@ -12,6 +12,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### [Unreleased]
```
Comment (chris-sanders) [inline charts/replicated-library/README_CHANGELOG.md.gotmpl:13]:
```suggestion

```

**Comment by chris-sanders** [comment]
Interesting looking at the values.yaml file this introduce several configurations that are specific to StatefulSets on the `app` object.

I think those are:
- serviceName
- persistentVolumeClaimTemplate
- something about the ClusterIP?

I'm just trying to figure out if this is the API we want for these. Having fields that only apply with another field is set, might be unavoidable but it feels strange.

For example, could you not define a volume under persistence and that just be used as a volumeClaimTemplate if the deployment it's mapped to is a Statefulset?

**Comment by chris-sanders** [comment]
Ok after discussing in person I think this is where we landed.

Allow statefulSet specific configs on the app so users have access.

persistentVolumClaimTemplate will need to be configured for a StatefulSet most of the time (probably what a user wants).

- We'll look at adding some storage abstraction so you can map a volume to the app and that would then handle choosing if you need a PVC (deployment) or PVCT (statefulset) so that it's consistent and the users doesn't have to understand that distinction to get started. This is expected to be non breaking so we can do that investigation at a later time.

serviceName is a statefulSet specific config, and for now it's just set to the app name. We should follow up with a future function that finds the service name and automatically sets it for you otherwise it's easy to misconfigure.

The clsuterIP setting was my misunderstanding, we'll update the phrasing to show that setting a service cluster IP to "None" is how you configure headless it's not unique to statefulSets.

File: charts/replicated-library/templates/lib/_statefulset.tpl [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  updateStrategy:
+    type: {{ $strategy }}
+    {{- if and (eq $strategy "RollingUpdate") $values.rollingUpdate.partition}}
+    rollingUpdate:
+      partition: {{ $values.rollingUpdate.partition }}
+    {{- end }}
+  selector:
+    matchLabels:
+      {{- include "replicated-library.labels.selectorLabels" . | nindent 6 }}
+  serviceName:  {{- default (include "replicated-library.names.appname" .) ($values.serviceName) }}
```
[the remark this answers is not in the document]
Comment (diamonwiggins) [reply charts/replicated-library/templates/lib/_statefulset.tpl:40]: I wouldn't be opposed, but it's only worth pointing out that it _technically_ isn't a required field as I think you can use a non headless service with a StatefulSet if you wanted to where you could just leave `serviceName` blank. I think most would use a headless service, but not sure if that's enough of a caveat where we would want it to not be required.

[the remark this answers is not in the document]
Comment (chris-sanders) [reply charts/replicated-library/templates/lib/_statefulset.tpl:40]: Oh that's a good point, I was thinking it was required but it's not. If you _do_ set this and it doesn't match the actual service name is that an error? Now I'm questioning if I understand the impact of us 'guessing wrong' in this case. Maybe this is actually fine.

File: charts/replicated-library/templates/lib/_statefulset.tpl [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    metadata:
```
Comment (chris-sanders) [inline charts/replicated-library/templates/lib/_statefulset.tpl:41]: This doesn't appear to be a valid format.
metadata is not a subfield of serviceName I think you've missed a key here somewhere.

File: charts/replicated-library/templates/lib/_statefulset.tpl [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        {{- . | nindent 8 }}
+      {{- end }}
+      labels:
+        {{- include "replicated-library.labels.selectorLabels" . | nindent 8 }}
+        {{- with $values.podLabels }}
+        {{- toYaml . | nindent 8 }}
+        {{- end }}
+    spec:
+      {{- include "replicated-library.pod" . | nindent 6 }}
+  volumeClaimTemplates:
```
Comment (chris-sanders) [inline charts/replicated-library/templates/lib/_statefulset.tpl:53]: I don't think volumeClaimTemplates are mandatory should this be behind an if statement?

File: charts/replicated-library/templates/lib/_statefulset.tpl (hunk 2 above)
Comment (chris-sanders) [inline charts/replicated-library/templates/lib/_statefulset.tpl:40]: I'll do further testing, but I think the leading "-" here is a problem. this is going to lave no space between the ":" and the value which isn't valid. I think.

I'm doing some testing I'll confirm this while I am.

**Review by chris-sanders — CHANGES_REQUESTED** [review]
Is this the correct branch, I see a few errors in here in particular the indention of the metadat field is missing the `template` key which as far as I can tell means this will never properly work as written.

**Comment by chris-sanders** [comment]
I've opened #31 to fix the errors listed above it was more involved than I could do directly on this PR. (The PR merges into this branch).

I think there's still a discussion to be had around the format of the persistentVolumeClaimTemplate api that was added, I've detailed that in #31

**Comment by chris-sanders** [comment]
I'm going to close this since #40 which was based on this landed.

**Pull request closed without merging**